### PR TITLE
[Issue 5562][docs]Modify schema compatibility check doc

### LINKED ISSUE #5562: Add the doc for new schema compatibility  strategy.
link #5227 , add the new schema compatibility strategy doc for 2.5.0 version in pulsar.

### PR DESCRIPTION
**Pull request by congbobo184**
Fixes #5562 
In order to add the doc for #5227 , to solve the issue #5562 
### Motivation
modify the doc.

### CODE REVIEW

**Comment by congbobo184** [comment]
@Jennifer88huang  @Anonymitaet  PTAL!

**Comment by Anonymitaet** [comment]
@congbobo184 I'll review later.

**Comment by congbobo184** [comment]
run cpp tests
run java8 tests

**Comment by congbobo184** [comment]
run integration/appveyor/pr tests

File: site2/docs/schema-evolution-compatibility.md [hunk 1]
```diff
@@ -366,6 +366,40 @@ Any order
...
+
+<td> 
+
+`UNDEFINED`
+
+</td> 
+
+<td> 
+
+The schema compatibility check strategy in order to Upgrade to 2.5.0 or above for compatibility with pulsar version below 2.5.0.
```
Comment (Anonymitaet) [inline site2/docs/schema-evolution-compatibility.md]:
```suggestion
* If you upgrade a lower version to Pulsar 2.5.0, then it uses the `UNDEFINED` strategy, at this time, it actually uses the compatibility strategy of the topic.

* If you use Pulsar 2.5.0 and a higher version, then it uses the `UNDEFINED` strategy, at this time, it actually uses the `FULL` compatibility strategy.
```

File: site2/docs/schema-evolution-compatibility.md [hunk 2]
```diff
@@ -366,6 +366,40 @@ Any order
...
+
+</td> 
+
+<td> 
+
+NONE
```
Comment (Anonymitaet) [inline site2/docs/schema-evolution-compatibility.md]:
```suggestion
It depends
```

File: site2/docs/schema-evolution-compatibility.md [hunk 3]
```diff
@@ -366,6 +366,40 @@ Any order
...
+
+</td> 
+
+<td> 
+
+NONE
+
+</td>
```
Comment (Anonymitaet) [inline site2/docs/schema-evolution-compatibility.md]:
```suggestion
It depends
```

File: site2/docs/schema-evolution-compatibility.md [hunk 4]
```diff
@@ -366,6 +366,40 @@ Any order
...
+
+<td> 
+
+NONE
+
+</td> 
+
+<td> 
+
+NONE
```
Comment (Anonymitaet) [inline site2/docs/schema-evolution-compatibility.md]:
```suggestion
It depends
```

File: site2/docs/schema-evolution-compatibility.md [hunk 5]
```diff
@@ -626,6 +660,44 @@ None
...
+
+<th>
+    
+Compatibility check strategy
+
+</th>
+
+<th>
+    
+Note
```
Comment (Anonymitaet) [inline site2/docs/schema-evolution-compatibility.md]:
```suggestion
Description
```

File: site2/docs/schema-evolution-compatibility.md [hunk 6]
```diff
@@ -626,6 +660,44 @@ None
...
+
+<td> 
+
+`UNDEFINED`
+
+</td> 
+
+<td> 
+
+The schema compatibility check strategy in order to Upgrade to 2.5.0 or above for compatibility with pulsar version below 2.5.0.
```
Comment (Anonymitaet) [inline site2/docs/schema-evolution-compatibility.md]:
```suggestion
* If you upgrade a lower version to Pulsar 2.5.0, then it uses the `UNDEFINED` strategy, at this time, it actually uses the compatibility strategy of the topic.

* If you use Pulsar 2.5.0 and a higher version, then it uses the `UNDEFINED` strategy, at this time, it actually uses the `FULL` compatibility strategy.
```

File: site2/docs/schema-evolution-compatibility.md [hunk 7]
```diff
@@ -781,3 +853,168 @@ Consequently, you can upgrade the producers and consumers in **any order**.
 </tr>
 
 </table>
+
+## Schema verification for producers
```
Comment (Anonymitaet) [inline site2/docs/schema-evolution-compatibility.md]: I've moved this part before `order of upgrading clients`, please delete lines 857-1017.

File: site2/docs/schema-evolution-compatibility.md [hunk 8]
```diff
@@ -626,6 +660,44 @@ None
...
+<td> 
+
+The schema compatibility check strategy in order to Upgrade to 2.5.0 or above for compatibility with pulsar version below 2.5.0.
+
+</td> 
+
+</tr>
+
+</table>
+
```
Comment (Anonymitaet) [inline site2/docs/schema-evolution-compatibility.md]: Add schema verification here. I've re-written some sentences as well.

```suggestion
## Schema verification

When a producer or a consumer tries to connect to a topic, a broker performs some checks to verify a schema.

### Producer

When a producer tries to connect to a topic (suppose ignore the schema auto creation), a broker does the following checks:

* Check if the schema carried by the producer exists in the schema registry or not.

    * If the schema is already registered, then the producer is connected to a broker and produce messages with that schema.

    * If the schema is not registered, then Pulsar verifies if the schema is allowed to be registered based on the configured compatibility check strategy.

### Consumer

When a consumer tries to connect to a topic, a broker checks if a carried schema is compatible with a registered schema based on the configured schema compatibility check strategy.

<table style="table">

<tr>

<th>

Compatibility check strategy

</th>

<th>

Check logic

</th>

</tr>

<tr>

<td> 

`ALWAYS_COMPATIBLE`

</td> 

<td> 

All pass

</td>  

</tr>

<tr>

<td> 

`ALWAYS_INCOMPATIBLE`

</td> 

<td> 

No pass

</td> 

</tr>

<tr>

<td> 

`BACKWARD`

</td> 

<td> 

Can read the last schema

</td> 

</tr>

<tr>

<td> 

`BACKWARD_TRANSITIVE`

</td> 

<td> 

Can read all schemas

</td> 

</tr>

<tr>

<td> 

`FORWARD`

</td> 

<td> 

Can read the last schema

</td> 

</tr>

<tr>

<td> 

`FORWARD_TRANSITIVE`

</td> 

<td> 

Can read the last schema

</td> 

</tr>

<tr>

<td> 

`FULL`

</td> 

<td> 

Can read the last schema

</td> 

</tr>

<tr>

<td> 

`FULL_TRANSITIVE`

</td> 

<td> 

Can read all schemas

</td> 

</tr>

</table>
```

**Comment by congbobo184** [comment]
@sijie 
```
bin/pulsar-admin namespaces set-schema-autoupdate-strategy --compatibility <compatibility-level> tenant/namespace
bin/pulsar-admin namespaces get-schema-autoupdate-strategy --compatibility <compatibility-level> tenant/namespace
```
I think this two command should not be provided to users.
It's easy to confuse users.
```
bin/pulsar-admin namespaces set-schema-compatibility-strategy --compatibility <compatibility-level> tenant/namespace
bin/pulsar-admin namespaces get-schema-compatibility-strategy --compatibility <compatibility-level> tenant/namespace
bin/pulsar-admin namespaces get-is-allow-auto-update-schema --disable tenant/namespace
```

**Comment by sijie** [comment]
@congbobo184 if those two commands are deprecated, you can add `hidden` flag to hide them.

**Comment by congbobo184** [comment]
This seems like a good way

File: site2/docs/schema-understand.md [hunk 9]
```diff
@@ -560,16 +563,10 @@ This diagram illustrates how does Schema work on the consumer side.
...
 2. The consumer connects to the broker with the `SchemaInfo` extracted from the passed-in schema instance.
    
-3. The broker looks up the schema in the schema storage to check if it is already a registered schema. 
-   
-4. If yes, the broker skips the schema validation since it is a known schema, and returns the schema version to the consumer.
-
-5. If no, the broker validates the schema based on the schema compatibility check strategy defined for the topic. 
-   
-6. If the schema is compatible, the broker stores it and returns the schema version to the consumer. 
+3. If the schema is compatible, the consumer will be connected.
```
Comment (Anonymitaet) [inline site2/docs/schema-understand.md]:
```suggestion
3. If the schema is compatible, the consumer is connected to the broker. 
```

File: site2/docs/schema-understand.md [hunk 10]
```diff
@@ -560,16 +563,10 @@ This diagram illustrates how does Schema work on the consumer side.
...
-   
-4. If yes, the broker skips the schema validation since it is a known schema, and returns the schema version to the consumer.
-
-5. If no, the broker validates the schema based on the schema compatibility check strategy defined for the topic. 
-   
-6. If the schema is compatible, the broker stores it and returns the schema version to the consumer. 
+3. If the schema is compatible, the consumer will be connected. 
    
-7. If the schema is incompatible, the consumer will be disconnected.
+4. If the schema is incompatible, the consumer will be disconnected.
```
Comment (Anonymitaet) [inline site2/docs/schema-understand.md]:
```suggestion
4. If the schema is incompatible, the consumer is disconnected to the broker.
```

**Comment by Anonymitaet** [comment]
@congbobo184 as we discussed before, we finalize contents first and then re-draw the workflow, so pls do not add images now.

**Comment by congbobo184** [comment]
@sijie  where does the flag add? pulsar-client-tools or broker reject this request? if it is in pular-client-tools, how to add it?

**Comment by sijie** [comment]
@congbobo184 this is one of the examples: https://github.com/apache/pulsar/blob/master/pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdNamespaces.java#L67

**Comment by congbobo184** [comment]
Thanks, i will look carefully and add hidden flag to hide them.  :)

File: site2/docs/schema-understand.md [hunk 11]
```diff
@@ -539,14 +539,17 @@ This diagram illustrates how does schema work on the Producer side.
 3. The broker looks up the schema in the schema storage to check if it is already a registered schema. 
    
 4. If yes, the broker skips the schema validation since it is a known schema, and returns the schema version to the producer.
+
+5. If no, the broker judge whether this namespace can automatically create schema.
+   isAllowAutoUpdateSchema can be modified by Admin-Api and REST-API 
   
-5. If no, the broker validates the schema based on the schema compatibility check strategy defined for the topic. 
+6. If yes, the broker validates the schema based on the schema compatibility check strategy defined for the topic.
```
[the remark this answers is not in the document]
Comment (congbobo184) [reply site2/docs/schema-understand.md]: yes, we can delete the line 546 👍 :)

File: site2/docs/schema-manage.md [hunk 12]
```diff
@@ -50,32 +58,47 @@ For a consumer, the `AutoUpdate` happens in the following cases:
...
+![AutoUpdate Consumer](assets/schema-autoupdate-consumer.png)
+        
 
 ### Manage AutoUpdate strategy
 
 You can use the `pulsar-admin` command to manage the `AutoUpdate` strategy as below:
 
+* [Enable AutoUpdate](#enable-autoupdate)
+
 * [Disable AutoUpdate](#disable-autoupdate)
```
Comment (Anonymitaet) [inline site2/docs/schema-manage.md]: As we discussed just now, please correct the commands for `enable` and `disable`

File: site2/docs/schema-understand.md [hunk 13]
```diff
@@ -539,8 +539,18 @@ This diagram illustrates how does schema work on the Producer side.
...
+  * If `isAllowAutoUpdateSchema` sets to **true**, then a schema can be created, and the broker validates the schema based on the schema compatibility check strategy defined for the topic.
+  
+  * If `isAllowAutoUpdateSchema` sets to **false**, then a schema can not be created, and the consumer is rejected to connect to the broker.
   
-5. If no, the broker validates the schema based on the schema compatibility check strategy defined for the topic. 
+**Tip**:
+
+`isAllowAutoUpdateSchema` can be set via **Pulsar admin API** or **REST API.** 
+
+For how to set `isAllowAutoUpdateSchema` via Pulsar admin API, see [Manage AutoUpdate Strategy](schema-manage.md/#manage-autoupdate-strategy).
```
Comment (Anonymitaet) [inline site2/docs/schema-understand.md]: Seems a little weird between #5 and #6? Do we need to add some contents like "If a schema is updated automatically:” after #5?

**Comment by congbobo184** [comment]
run cpp tests

**Comment by congbobo184** [comment]
run Integration Tests

**Comment by Jennifer88huang** [comment]
@congbobo184 @Anonymitaet When you fix an issue, please add the issue number and component in the title, for example, "[Issue 5562][docs] xxx", and add "Fixes #xxxx" in the description.
So when the PR is merged, the corresponding issue will be closed automatically.

**Comment by sijie** [comment]
> @congbobo184 @Anonymitaet When you fix an issue, please add the issue number and component in the title, for example, "[Issue 5562][docs] xxx", and add "Fixes #xxxx" in the description.
> So when the PR is merged, the corresponding issue will be closed automatically.

maybe we can add this to the contributor guide and encourage committers and contributors following the same practice.

**Pull request merged**